app/controllers/flood_risk_engine/enrollments/steps_controller.rb

```diff
@@ -8,30 +8,81 @@
 #
 module FloodRiskEngine
   module Enrollments
-    class StepsController < ApplicationController
-      class StepError < StandardError; end
+    class StepsController < ApplicationController # rubocop:disable Metrics/ClassLength
+
       rescue_from StepError, with: :step_not_found
+
       before_action :check_step_is_valid
       before_action :back_button_cache_buster
+      before_action :clear_error_params, only: [:update, :process_postcode, :process_address]
 
       def show
         form.validate(session[:error_params]) if params[:check_for_error]
+
+        form.prepopulator(params) if form.respond_to?(:prepopulator)
+
         render :show, locals: locals
       end
 
       def update
-        clear_error_params
         success = save_form!
-        if form.redirect?
-          redirect_to(form.redirection_url)
-        elsif success
-          step_forward
-          redirect_to step_url
+
+        # The redirect was happening regardless of outcome, hence moved to within success
+        # i.e if validate or save fails we should NOT redirect
+        if success
+
+          if form.redirect?
+            redirect_to(form.redirection_url)
+          else
+            step_forward
+            redirect_to step_url
+          end
+
         else
           handle_failure_and_redirect_back_to_show
         end
       end
 
+      def process_postcode # rubocop:disable Metrics/AbcSize
+        # We redirect here from update when dealing with page containing Postcode entry and form validated
+        enrollment.go_forward
+        enrollment.save
+
+        # form & step cache results so don't return what we need after go_forward - so this ensures the right values
+        address_form = Steps::FormObjectFactory.form_object_for(enrollment.step, enrollment)
+
+        # Check if address form needs prepopulatingm, for example, so the view can be populated with address list
+        address_form.prepopulator(params) if address_form.respond_to?(:prepopulator)
+
+        locals_after_redirect = {
+          form: address_form,
+          step: enrollment.step
+        }
+
+        logger.info("Rendering step #{enrollment.step} for Postcode [#{params[:post_code]}]")
+
+        render :show, locals: locals_after_redirect
+      end
+
+      def process_address
```

Comment:
I can't recall all the issues I had with your generic methods now , but I tried alot of things, and this  was the way I got everything working and clean behaviour for back buttons .

When the manual entry element is merged I can revisit these.